### DISCUSSION TITLE
[RFC] 031 - URL 分享设置参数

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#1632] [RFCs]
## 背景

- #1059
- #917

## 设计思路

我期望该方案可以考虑更多的用户场景，结合 #913 做的思路，我认为可以参考此模式，实现 URL 层的设置参数分享与导入。

### schema定义

设计 settings 层允许分享的配置schema，比如 OpenAI 的 key 与 proxy URL ，再比如是否开启 openrouter 的provider 等。这部分可以参考现有的 settings层类型定义 https://github.com/lobehub/lobe-chat/blob/aaedfa716ea6d0a48266f0572d811729c7a0a25a/src/types/settings/index.ts#L34-L39 
个人建议 searchParams 的 key 可以使用 `settingLanguageModel` 这种。

### 触发分享

1. 在设置页面提供分享按钮，点击后弹出模态框，允许用户配置需要分享的设置，允许只分享部分参数，并实时展示对应的 url 参数。
2. 在 shareService 中实现分享 url 逻辑。

### 导入分享链接逻辑

目前 settingsStore 仍然存在一个 preference 的水合逻辑  ，因此直接useEffect 读取数据会有问题，需要使用 `useEffectAfterHydrated` 。 (未来 preference 的水合逻辑会被移除，但现在临时还需要这个) https://github.com/lobehub/lobe-chat/blob/7a51329bc40bf936964399f41cda56cfe5479a05/src/layout/GlobalLayout/StoreHydration.tsx#L17-L20
可以使用 nextjs 的 `useSearchParams` 来读取对应的 searchParams  https://nextjs.org/docs/app/api-reference/functions/use-search-params ，然后在 useEffectAfterHydrated 中 setSettings 到 store 中

**Comment by cy948** [comment]
所以说设计思路大概是这样吗？
- URL 参数规范：操作格式参照 [json patch](https://datatracker.ietf.org/doc/html/rfc6902) ，参数命名参照 [settings schema](https://github.com/lobehub/lobe-chat/blob/aaedfa716ea6d0a48266f0572d811729c7a0a25a/src/types/settings/index.ts#L34-L39)
- URL导入设置：使用`useSearchParams`hook读取外部传入字符串 -> 使用 [jiff](https://www.npmjs.com/package/jiff) 库解析 -> 在[水合逻辑](https://github.com/lobehub/lobe-chat/blob/7a51329bc40bf936964399f41cda56cfe5479a05/src/layout/GlobalLayout/StoreHydration.tsx#L17-L20)中写入设置

**Reply by arvinxx** [reply]
不需要用json patch，直接读 settings就好，最后我没有用 json patch 的实现

**Reply by cy948** [reply]
按照这种格式？
```txt
model=gpt-4-1106-preview;params.max_tokens=300;
```

**Reply by arvinxx** [reply]
看实现成本吧，我感觉可以考虑直接用 json 转义。上述这个格式只是为了写起来方便，url这个可以由程序生成

**Reply by cy948** [reply]
同意
